--- a/wemo_ctrl/wemo_ctrl.c
+++ b/wemo_ctrl/wemo_ctrl.c
@@ -2493,7 +2493,7 @@ int wemoCtrlPointStart(char *ifname, pri
     LOG_INFO_MSG("Initializing UPnP Sdk with ifname = %s port = %u",
                       ifname, port);
 
-    rc = UpnpInit2(ifname, port, DeviceUDN);
+    rc = UpnpInit2(ifname, port);
     if (rc != UPNP_E_SUCCESS) {
         LOG_ERROR_MSG("UpnpInit2() Error: %d (%s)", rc, UpnpGetErrorMessage(rc));
         UpnpFinish();
